### PR TITLE
Command autocomplete

### LINKED ISSUE #136: Command preview

### PR DESCRIPTION
**Pull request by louislai**
##Resolve #136 #141 

Check out the preview for more details

untitled

Command preview will auto-refresh whenever u type a new text to suggest the command that you most likely wanna type.
Scenarios:
- Only one suggested command: Pressing TAB will autocomplete your command
- Multiple suggested commands: PRessing TAB will cycle thru the different suggestions. Pressing ENTER while highlighting a suggestion will choose that suggestion
- Zero suitable suggested command: The command is invalid. There are visual hints given

Even preview aliased command as well!

Do comment on the UX

### CODE REVIEW

File: src/main/java/seedu/toluist/commons/util/StringUtil.java [hunk 1]
```diff
@@ -101,4 +101,29 @@ public static String nounWithCount(String noun, int count) {
...
     }
+
+    /**
+     * Return a transformed string of an original string, with the first word replaced
+     * @param s the original string
+     * @param replacement the string to replace the first word
+     * @return the transformed string
+     */
+    public static String replaceFirstWord(String s, String replacement) {
+        if (replacement == null) {
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/commons/util/StringUtil.java]: if (replacement == null || !StringUtil.isPresent(s))) ?

File: src/main/java/seedu/toluist/dispatcher/CommandDispatcher.java [hunk 2]
```diff
@@ -60,7 +59,7 @@ public void dispatchRecordingHistory(String command) {
     }
 
     public void dispatch(String command) {
-        String deAliasedCommand = aliasConfig.dealias(command);
+        String deAliasedCommand = getDealiasedCommand(command);
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/dispatcher/CommandDispatcher.java:66]: might wanna keep it consistent on `deAliased` vs `Dealiased`

File: src/main/java/seedu/toluist/ui/UiStore.java [hunk 3]
```diff
@@ -24,6 +26,8 @@
  * Ui will observe for changes to `shownTasks` and re-render on state change
  */
 public class UiStore {
+    public static final int INVALID_SUGGESTION_INDEX = -1;
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/ui/UiStore.java]: Im not sure about this cuz of my sucky english so I'll just flag this first.
Should it be INDEX_INVALID_SUGGESTION or INVALID_SUGGESTION_INDEX based on the pattern CATEGORY_DETAIL?

File: src/main/java/seedu/toluist/ui/UiStore.java [hunk 4]
```diff
@@ -78,6 +85,15 @@ public void setCommandResult(CommandResult commandResult) {
         return observableCommandText;
     }
 
+    public void setSuggestedCommands(List<String> suggestedCommands) {
+        observableSuggestedCommandIndex.set(-1);
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/ui/UiStore.java]: .set(INVALID_SUGGESTION_INDEX)

**Review by Melvin-Tan — COMMENTED** [review]
This drops our test coverage quite a bit hahahaha

**Comment by Melvin-Tan** [comment]
screen shot 2017-03-25 at 4 03 33 am
There's a bug here when u try to increase the height of the app. (I checked master branch and it's fine)

File: src/main/java/seedu/toluist/ui/UiStore.java (hunk 3 above)
Comment (louislai) [reply src/main/java/seedu/toluist/ui/UiStore.java]: screen shot 2017-03-25 at 09 12 59
not sure actually

**Comment by louislai** [comment]
Okay let's say `add` is aliased to `ravl`. Should I show `add` when `r` is typed?

**Comment by Melvin-Tan** [comment]
Current implementation supports that right, i tried alias mc mark complete, when i type mc, mark complete appears. i think that is intended and is good.

**Comment by louislai** [comment]
@Melvin-Tan let's say i alias `mark complete` to `cvg`. right now `mark complete` only shows up as an option if i type `cvg`. Should i show it immediately when i type `c`?

**Comment by Melvin-Tan** [comment]
Yeah why not?

**Comment by louislai** [comment]
@Rinder5 ?

**Comment by louislai** [comment]
@Melvin-Tan handled

**Review by Melvin-Tan — APPROVED** [review]

**Comment by Melvin-Tan** [comment]
the new UI might take some time for me to get used to lol

**Pull request merged**